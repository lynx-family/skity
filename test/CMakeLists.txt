# Copyright 2021 The Lynx Authors. All rights reserved.
# Licensed under the Apache License Version 2.0 that can be found in the
# LICENSE file in the root directory of this source tree.

add_subdirectory(${CMAKE_SOURCE_DIR}/third_party/gtest third_party/gtest)

add_subdirectory(ut)

if(${SKITY_SW_RENDERER})
  add_subdirectory(bench)
endif()

if (${SKITY_MTL_BACKEND} AND ${SKITY_CODEC_MODULE})
  add_subdirectory(golden)
endif()

