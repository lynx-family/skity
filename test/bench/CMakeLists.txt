# Copyright 2021 The Lynx Authors. All rights reserved.
# Licensed under the Apache License Version 2.0 that can be found in the
# LICENSE file in the root directory of this source tree.

add_subdirectory(${CMAKE_SOURCE_DIR}/third_party/google_benchmark third_party/google_benchmark)
add_subdirectory(micro)

if (NOT APPLE)
    return()
endif()

add_executable(skity_hw_render_bench
    hw_benchmarks.cc
    case/draw_circle.cc
    case/draw_circle.hpp
)

target_sources(skity_hw_render_bench PRIVATE 
    common/bench_target.cc
    common/bench_target.hpp
    common/bench_context.cc
    common/bench_context.hpp
)

target_compile_options(skity_hw_render_bench PRIVATE "-fobjc-arc")

if (${SKITY_MTL_BACKEND})
  target_compile_definitions(skity_hw_render_bench PRIVATE -DSKITY_BENCH_MTL_BACKEND=1)
  target_sources(skity_hw_render_bench PRIVATE 
    common/bench_target_mtl.h
    common/bench_target_mtl.mm
    common/bench_context_mtl.h
    common/bench_context_mtl.mm
  )
  target_link_libraries(skity_hw_render_bench PRIVATE "-framework Metal")
endif()

if (${SKITY_GL_BACKEND})
    target_compile_definitions(skity_hw_render_bench PRIVATE -DSKITY_BENCH_GL_BACKEND=1)
    target_sources(skity_hw_render_bench PRIVATE 
        common/bench_target_gl.cc
        common/bench_target_gl.hpp
        common/bench_context_gl.cc
        common/bench_context_gl.hpp
        common/bench_gl_context_mac.h
        common/bench_gl_context_mac.mm
        common/bench_gl_context.hpp
    )
    target_link_libraries(skity_hw_render_bench PRIVATE "-framework OpenGL")
endif()

target_include_directories(skity_hw_render_bench
    PUBLIC
    ${CMAKE_SOURCE_DIR}
)

target_compile_options(skity_hw_render_bench PUBLIC -fno-rtti)
target_compile_options(skity_hw_render_bench PUBLIC -std=c++17)
target_compile_definitions(skity_hw_render_bench PUBLIC -DDISABLE_SKITY_EXPERIMENTAL_WARNINGS)

target_link_libraries(skity_hw_render_bench
    PUBLIC
    skity::codec
    skity::skity
    benchmark::benchmark
)
target_link_libraries(skity_hw_render_bench
    PRIVATE "-framework Carbon"
            "-framework AppKit")

target_link_libraries(skity_hw_render_bench PRIVATE glm::glm-header-only)
