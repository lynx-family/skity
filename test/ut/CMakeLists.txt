# Copyright 2021 The Lynx Authors. All rights reserved.
# Licensed under the Apache License Version 2.0 that can be found in the
# LICENSE file in the root directory of this source tree.

include(GoogleTest)

enable_testing()

# Test case list
add_executable(skity_unit_test
    effect/image_filter_test.cc
    geometry/geometry_test.cc
    geometry/math_test.cc
    geometry/matrix_test.cc
    geometry/quaternion_test.cc
    geometry/rect_test.cc
    geometry/vector_test.cc
    graphic/path_measure_test.cc
    graphic/path_test.cc
    render/canvas_state_test.cc
    render/resource_cache_test.cc
    recorder/display_list_test.cc
    text/text_test.cc
    utils/arena_allocator_test.cc
    utils/array_list_test.cc
)

target_include_directories(skity_unit_test
    PUBLIC
    ${CMAKE_SOURCE_DIR}
)

target_link_libraries(skity_unit_test
    PUBLIC
    skity::skity
    gmock_main
)

target_link_libraries(skity_unit_test PRIVATE glm::glm-header-only)

gtest_add_tests(skity_unit_test "" AUTO)
