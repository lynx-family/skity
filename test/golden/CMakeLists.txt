# Copyright 2021 The Lynx Authors. All rights reserved.
# Licensed under the Apache License Version 2.0 that can be found in the
# LICENSE file in the root directory of this source tree.

option(SKITY_GOLDEN_GUI "Option to open GUI for golden test" ON)

if (NOT APPLE)
  message(FATAL_ERROR "Currently golden test only support Metal environment")
endif()

if (NOT ${SKITY_MTL_RENDERER})
  message(FATAL_ERROR "Currently golden test only support Metal renderer")
endif()

set(CMAKE_OBJCXX_FLAGS "${CMAKE_OBJCXX_FLAGS} -fobjc-arc", "-std=c++17")

add_executable(skity_golden_test
  ${CMAKE_CURRENT_LIST_DIR}/golden_test.cc
  ${CMAKE_CURRENT_LIST_DIR}/common/golden_test_check.cc
  ${CMAKE_CURRENT_LIST_DIR}/common/golden_test_check.hpp
  ${CMAKE_CURRENT_LIST_DIR}/common/golden_test_env.cc
  ${CMAKE_CURRENT_LIST_DIR}/common/golden_test_env.hpp
  ${CMAKE_CURRENT_LIST_DIR}/common/golden_texture.cc
  ${CMAKE_CURRENT_LIST_DIR}/common/golden_texture.hpp

  # Metal backend code
  ${CMAKE_CURRENT_LIST_DIR}/common/mtl/golden_test_env_mtl.h
  ${CMAKE_CURRENT_LIST_DIR}/common/mtl/golden_test_env_mtl.mm
  ${CMAKE_CURRENT_LIST_DIR}/common/mtl/golden_texture_mtl.h
  ${CMAKE_CURRENT_LIST_DIR}/common/mtl/golden_texture_mtl.mm
)

target_include_directories(skity_golden_test
  PUBLIC
  ${CMAKE_CURRENT_LIST_DIR}
)

target_link_libraries(skity_golden_test
  gtest
  skity::skity
  skity::codec
  "-framework Metal"
  "-framework CoreGraphics"
  "-framework ImageIO"
  "-framework CoreServices"
)

target_compile_definitions(skity_golden_test PUBLIC -DDISABLE_SKITY_EXPERIMENTAL_WARNINGS)


if (${SKITY_GOLDEN_GUI})
  add_subdirectory(playground)
endif()

set(SKITY_GOLDEN_TEST_DIR ${CMAKE_CURRENT_LIST_DIR})

### add golden test case to skity_golden_test
###
### @param NAME test case name
### @param SOURCES test case source files
### @param INCLUDES test case include directories
### @param LIBS test case link libraries
### @param DEFINES test case compile definitions
###
### @example
### add_golden_test(
###   NAME      <case name>
###   SOURCES   <source files>
###   INCLUDES  <custom include path if needed>
###   LIBS      <other library if needs>
###   DEFINES   <other compile definitions>
### )
###
function(add_golden_test)
  set(options )
  set(oneValueArgs NAME)
  set(multiValueArgs SOURCES INCLUDES LIBS DEFINES)

  cmake_parse_arguments(
    PARSE_ARGV 0
    ARG
    "${options}"
    "${oneValueArgs}"
    "${multiValueArgs}"
  )

  if (NOT ARG_NAME)
    message(FATAL_ERROR "add_golden_test: NAME is required")
  endif()

  message(STATUS "add_golden_test: ${ARG_NAME}")

  if (NOT ARG_SOURCES)
    message(FATAL_ERROR "add_golden_test: SOURCES is required")
  endif()

  add_library(${ARG_NAME} OBJECT
    ${ARG_SOURCES}
  )

  if (ARG_INCLUDES)
    target_include_directories(${ARG_NAME}
      PRIVATE
      ${ARG_INCLUDES}
    )
  endif()

  target_link_libraries(${ARG_NAME} PRIVATE gtest skity::skity)
  target_include_directories(${ARG_NAME} PRIVATE ${SKITY_GOLDEN_TEST_DIR})

  if (ARG_LIBS)
    target_link_libraries(${ARG_NAME}
      ${ARG_LIBS}
    )
  endif()

  target_compile_definitions(${ARG_NAME}
    PRIVATE
    CASE_NAME="${ARG_NAME}"
    CASE_DIR="${CMAKE_CURRENT_LIST_DIR}"
  )

  # append other user defined defines
  if (ARG_DEFINES)
    target_compile_definitions(${ARG_NAME}
      PRIVATE
      ${ARG_DEFINES}
    )
  endif()

  target_link_libraries(skity_golden_test
    $<TARGET_OBJECTS:${ARG_NAME}>
  )

endfunction(add_golden_test)

add_subdirectory(cases)
